### PR TITLE
Correctly control the state of a JamSession

### LINKED ISSUE #82: Correctly control the state of a JamSession

### PR DESCRIPTION
**Pull request by exp625**
Fixes #82. Also:
- Add more Websocket events
- 💥 Change method for the queue/delete entpoint

### CODE REVIEW

File: api/server/routes.go [hunk 1]
```diff
@@ -72,7 +72,7 @@ func (s *Server) registerQueueRoutes(r *mux.Router) {
 	r.Handle(queueIndexPath, s.jamSessionRequired(http.HandlerFunc(s.getQueue))).Methods("GET")
 	r.Handle(queueCollectionPath, s.jamSessionRequired(s.hostRequired(http.HandlerFunc(s.addCollection)))).Methods("PUT")
 	r.Handle(queueVotePath, s.jamSessionRequired(http.HandlerFunc(s.vote))).Methods("PUT")
-	r.Handle(queueDeletePath, s.jamSessionRequired(s.hostRequired(http.HandlerFunc(s.deleteSong)))).Methods("DELETE")
+	r.Handle(queueDeletePath, s.jamSessionRequired(s.hostRequired(http.HandlerFunc(s.deleteSong)))).Methods("PUT")
```
Comment (DrJoLe) [inline api/server/routes.go:75]: Why is this necessary? Since we are not conforming to REST principles anyways, this change is redundant.

File: docs/documentation.md [hunk 2]
```diff
@@ -107,13 +107,16 @@ The JamSession state will determine if the conductor has the right to control th
...
+| event                                                  | change to       
 |----------	                                             |-----------------
-| *User* creates a *JamSession*                          | active = ``true`` if the user has a selected Playback Device, else inactive = ``false`` 
-| *User* pauses playback through the *JamFactory App*    | inactive = ``false``
-| *User* resumes playback through the *JamFactory App*   | active = ``true``
-| *User* pauses playback through *Spotify*               | inactive = ``false``
-| *User* resumes/starts playback through *Spotify*       | inactive = ``false``
+| *User* creates a *JamSession*                          | inactive
+| *User* sets the *JamSession to active                  | active
+| *User* sets the *JamSession to inactive                | inactive
```
Comment (DrJoLe) [inline docs/documentation.md:114]: There are missing asterisks around `JamSession`.

File: docs/documentation.md [hunk 3]
```diff
@@ -521,7 +524,7 @@ Requires the user to be the host of the JamSession.
 ***Endpoint:***
 
 ```bash
-Method: DELETE
+Method: PUT
```
Comment (DrJoLe) [inline docs/documentation.md:527]: See above.

File: pkg/jamsession/jamsession.go [hunk 4]
```diff
@@ -72,11 +73,15 @@ type JamSession interface {
...
-	SetDeviceID(spotify.ID)
-	// SetDevice TODO
-	SetDevice(id string) error
+	SetDevice(string) error
+	// SocketJamUpdate updates the JamSession through the Websocket connection
+	SocketJamUpdate()
+	// SocketQueueUpdate updates the Queue through the Websocket connection
+	SocketQueueUpdate()
+	// SocketPlaybackUpdate updates the Playback through the Websocket connection
+	SocketPlaybackUpdate()
```
Comment (DrJoLe) [inline pkg/jamsession/jamsession.go:84]: We should think about what methods we actually need in the interface and what methods are specific to the Spotify implementations. No changes required here for now though.

**Review by DrJoLe — CHANGES_REQUESTED** [review]

File: api/server/routes.go (hunk 1 above)
Comment (exp625) [reply api/server/routes.go:75]: We don't, but the Angular http service for example. It does not support delete requests with a body. Also, the delete method was a misfit. We should stick to get and put requests only.

Maybe one day well change to /queue/jamlabel/spotifyid. But atm I don't see an advantage over this method

File: docs/documentation.md (hunk 2 above)
Comment (exp625) [reply docs/documentation.md]: Fixed with 449a8779ee0b8a2f92216ed05b8c006eebb72fb9

File: docs/documentation.md (hunk 3 above)
Comment (exp625) [reply docs/documentation.md:527]: See comment above

File: pkg/jamsession/jamsession.go (hunk 4 above)
Comment (exp625) [reply pkg/jamsession/jamsession.go:84]: Yes, the interface currently is a mess. But I think it's still better to repeat the same code multiple times to send the same websocket message on different occasions

File: api/server/routes.go (hunk 1 above)
Comment (DrJoLe) [reply api/server/routes.go:75]: The Angular http client actually does support using bodies on DELETE requests:

https://stackoverflow.com/a/40857437

I would prefer not to change this here, since I would like the API to be at least some kind of stable right now. I would accept the change here, if you are willing to fix the new API call in jamfactory-web as well as in the secret thing.

I don't want to blindly block the change but want you to know the impact it would have. So if you want to fix all places where this change would have an impact, go ahead and do so.

File: pkg/jamsession/jamsession.go (hunk 4 above)
Comment (DrJoLe) [reply pkg/jamsession/jamsession.go:84]: I agree on both statements.

File: api/server/routes.go (hunk 1 above)
Comment (exp625) [reply api/server/routes.go]: Reverted the change for now. On the next occasion we should clean up the method types.

Comment (exp625) [reply api/server/routes.go]: And why is the flutter app such a big secret?

Comment (DrJoLe) [reply api/server/routes.go]: Because the repository is still private
